src/core/tactics.ml

```diff
@@ -12,11 +12,38 @@ open Print
 let log_tact = new_logger 't' "tact" "tactics"
 let log_tact = log_tact.logger
 
+(* Helper functions related to goals *)
+let is_goal_typ = function Goal.GoalTyp _ -> true  | Goal.GoalUnif _ -> false
+let is_unif_typ x = not (is_goal_typ x)
+let goal_unif_of = fun cs -> Goal.GoalUnif cs
+let goal_typ_of  = fun x  -> Goal.GoalTyp x
+let cs_of_goal_unif = function
+  | Goal.GoalUnif cs -> cs
+  | _ -> assert false
+let g_of_goal_typ = function
+  | Goal.GoalTyp g -> g
+  | _ -> assert false
+
+(** [solve ps pos] calls the default solve algorithm on the unification
+    goals of the proof state [ps] and does not fail *)
+let solve ps pos =
+  try
+    let gs_typ,gs_unif = List.partition is_goal_typ ps.proof_goals in
+    let to_solve = List.map cs_of_goal_unif gs_unif in
+    let new_cs = Unif.solve {Unif.empty_problem with to_solve} in
+    let new_gs_unif = List.map goal_unif_of new_cs in
+    {ps with proof_goals = new_gs_unif @ gs_typ}
+  with
+  | Unif.Unsolvable ->
+    wrn pos "Unsolvable, try another unification tactic";
```

Comment:
Trying another tactic won't change anything when a constraint is unsolvable. The user should backtrack instead if that's possible. Otherwise the theorem or the definition is definitely wrong or untypable.

Reply:
replaced by fatal in ed44999